distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso

```diff
@@ -1169,3 +1189,113 @@ type Incomparable_Values_Error
 handle_incomparable_value ~function =
     handle t = Panic.catch t handler=(Error.throw Incomparable_Values_Error)
     handle No_Such_Method_Error <| handle Type_Error <| handle Unsupported_Argument_Types <| function
+
+## PRIVATE
+   Resolves a vector of ranges or indices into a vector of ranges that fit
+   within a sequence.
+resolve_ranges : Vector (Integer | Range) -> Integer -> Vector Range
+resolve_ranges ranges length =
+    ## Ensures that a descriptor fits within the range of the current
+       vector, trimming it or reporting an error if it is invalid.
+    trim descriptor = case descriptor of
+        Integer ->
+            actual_index = if descriptor < 0 then length + descriptor else descriptor
+            if (actual_index < 0) || (actual_index >= length) then Panic.throw (Index_Out_Of_Bounds_Error descriptor length) else
+                actual_index
+        Range start end step ->
+            if step <= 0 then Panic.throw (Illegal_Argument_Error "Range step must be positive.") else
+                if (start < 0) || (end < 0) then Panic.throw (Illegal_Argument_Error "Range start and end must not be negative.") else
+                    if start >= length then Panic.throw (Index_Out_Of_Bounds_Error start length) else
+                        actual_end = Math.min end length
+                        if actual_end < start then Range start start step else
+                            Range start actual_end step
+    ranges.map trim
+
+## PRIVATE
+   Creates a new vector where for each range, a corresponding section of the
+   source vector is added to the result.
+
+   Assumes that the ranges have been already bounds-checked (for example by
+   passing them through `resolve_ranges`).
+slice_ranges vector ranges =
+    if ranges.length == 0 then empty else case ranges.length == 1 of
+        True -> case ranges.first of
```

Comment:
Think it reads cleaner like:
```
slice_ranges vector ranges =
    if ranges.length == 0 then empty else
        if ranges.length != 1 then slice_many_ranges vector ranges else
            case ranges.first of
                Integer -> [vector.unsafe_at ranges.first]
                Range start end step -> case step == 1 of
                    True -> vector.slice start end
                    False -> slice_many_ranges vector ranges
```

Reply:
I wanted to avoid so much indentation, but other than that you are right. I will change it and hope we get multiline `if` soon-ish.